### PR TITLE
ION-5684 cloudhub insights and replay for 3.8

### PR DESCRIPTION
**Pull request by xuanshi**

### CODE REVIEW

File: mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/FlowSourceCloudhubInternalHandler.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+@Singleton
+@Named("mule.agent.tracking.handler.cloudhub.source")
+public class FlowSourceCloudhubInternalHandler extends BufferedHandler<FlowSourceEvent> {
+
+    private static final Logger LOGGER = LogManager.getLogger(FlowSourceCloudhubInternalHandler.class);
+
+    @Inject
+    @Named("mule.agent.tracking.replay.cloudhub.s3")
+    private ReplayStore replayStore;
```
Comment (wpoch) [inline mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/FlowSourceCloudhubInternalHandler.java]: I found the only usage is on the `flush` method and is casted to `CloudhubS3ReplayStore` maybe we can just inject that type and avoid the `named` and `instanceof` check.

File: mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/FlowSourceCloudhubInternalHandler.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    protected boolean flush(Collection<FlowSourceEvent> messages) {
+        LOGGER.debug("Flushing tracking payloads");
+        try {
+            ((CloudhubS3ReplayStore) replayStore).putFlowSourceEvent(messages);
+            return true;
+        } catch (Exception e) {
+            LOGGER.warn("Cloud not send payload to s3");
```
Comment (wpoch) [inline mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/FlowSourceCloudhubInternalHandler.java]: Please change it to 
```
LOGGER.warn("Cloud not send payload to s3. Error: {}", ExceptionUtils.getRootCauseMessage(ex));
LOGGER.debug(ex);
````

File: mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/CloudhubS3ReplayStore.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            event.setApplicationName(appName);
+            event.setFlowName(flowName);
+            event.setMessageId(transactionId);
+        }
+        return event;
+    }
+
+    void putFlowSourceEvent(Collection<FlowSourceEvent> events) {
+        for (FlowSourceEvent event : events) {
+            putFlowSourceEvent(event);
```
Comment (wpoch) [inline mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/CloudhubS3ReplayStore.java]: Please add a 
```
try {
   putFlowSourceEvent(event);
} catch (Exception ex) {
   throw new S3StorageException(format("Error storing the event: app: %s - flow: %s - message: %s", evet.getAppName(), app.getFlowName(), event.getMessageId(), ), ex);
}

File: mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/test/resources/log4j2.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (xuanshi) [inline mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/test/resources/log4j2.xml]: it's convenient to see logs while writing tests. I'll remove this

File: mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/assembly.xml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<assembly>
```
Comment (xuanshi) [inline mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/assembly.xml]: I was following the mule-agent-analytics-event-tracker. I'll remove this.

File: mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/CloudhubS3Configuration.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package com.mulesoft.agent.eventtracker.cloudhub.source;
+
+class CloudhubS3Configuration {
+
+    private String encryptionKey = "861c3a34-92e8-4392-9d41-69fa6bf09912";
```
Comment (xuanshi) [inline mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/CloudhubS3Configuration.java]: Is there a way to set default values to this? The hard coded value won't be visible anyway, I'll remove this then. It's for s3 object encryption.

File: mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/CloudhubS3ReplayStore.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    private String bucket;
+
+    private S3FlowSourceNameFactory nameFactory;
+
+    @PostConfigure
+    public void init() throws InitializationException {
+        AWSEncryptedKeyProvider keyProvider = new AWSEncryptedKeyProvider(cloudhubS3.getEncryptionKey());
+        String accessKey = keyProvider.getAccessKey();
+        String secretKey = keyProvider.getSecretKey();
+        bucket = System.getProperty(S3_BUCKET_KEY);
```
Comment (xuanshi) [inline mule-agent-eventtracking-internalhandlers/mule-agent-cloudhub-event-tracker/src/main/java/com/mulesoft/agent/eventtracker/cloudhub/source/CloudhubS3ReplayStore.java]: actually no. The bucket will be different between different workers. Cloudhub injects the bucket as a JVM property.

**Comment by wpoch** [comment]
@xuanshi I will review this PR tomorrow, please when we are providing feedback don't rewrite the branch history using a `push --force` so I can see the delta between changes, and is easier for me :)

**Comment by jeff-1amstudios** [comment]
@wpoch when you have some time can we merge this one?

**Comment by wpoch** [comment]
@jeff-1amstudios I spoke with @xuanshi about changing the way we work. Since those modules won't be bundled with the Mule Agent Plugin and are CH specific, maybe makes sense to just drop the compiled jars on the AMI `$MULE_HOME/plugins/mule-agent-plugin/lib/modules/` that CH uses, and the source code will be on another repo managed and maintained by the CH team.
This will allow everyone to get features faster and don't depend on each other.
Also, this way of work will enable us to fix issues quicker on production.
Let me know what do you think about this.

**Comment by jeff-1amstudios** [comment]
@wpoch perfect, love it :)

**Pull request closed without merging**